### PR TITLE
Package download request resuming

### PR DESCRIPTION
**Pull request by tmeasday**
For #7267 

I wanted to get your thoughts on this approach @benjamn.

Todo:

 - [x] Ensure it actually works with package downloading
 - [x] Figure out if there's a way to make the test meaningful
 - [x] Add a timeout to the retries
 - [x] Better logging
 - [x] Add `stream-buffers` to the dev bundle

### CODE REVIEW

File: tools/utils/http-helpers.js [hunk 1]
```diff
@@ -358,6 +369,47 @@ _.extend(exports, {
...
+    const MAX_ATTEMPTS = 5;
+    function attempt(triesRemaining) {
+      const headers = _.extend({}, urlOrOptions.headers, {
+        range: `bytes=${outputStream.size()}-`,
+      });
+      try {
+        return httpHelpers.request({
+          outputStream,
+          headers,
+          ...urlOrOptions,
```
Comment (tmeasday) [inline tools/utils/http-helpers.js]: XXX: should be `options`

File: tools/utils/http-helpers.js [hunk 2]
```diff
@@ -358,6 +369,47 @@ _.extend(exports, {
...
+      url: urlOrOptions,
+    };
+
+    const outputStream = new WritableStreamBuffer();
+
+    const MAX_ATTEMPTS = 5;
+    function attempt(triesRemaining) {
+      const headers = _.extend({}, urlOrOptions.headers, {
+        range: `bytes=${outputStream.size()}-`,
+      });
```
Comment (benjamn) [inline tools/utils/http-helpers.js]:
```js
const headers = {
  ...urlOrOptions.headers,
  range: `bytes=${outputStream.size()}-`
};
```

File: tools/utils/http-helpers.js [hunk 3]
```diff
@@ -358,6 +369,47 @@ _.extend(exports, {
...
+        return httpHelpers.request({
+          outputStream,
+          headers,
+          ...urlOrOptions,
+        });
+      } catch (e) {
+        console.log(e);
+        if (triesRemaining > 0) {
+          console.log(`Request failed, retrying`);
+          return attempt(triesRemaining - 1);
```
Comment (benjamn) [inline tools/utils/http-helpers.js]: What if the failure only counted against `triesRemaining` if the download made no progress (or less than some small amount of progress)?

File: tools/tests/utils-tests.js [hunk 4]
```diff
@@ -152,3 +152,30 @@ selftest.define("parse url", function () {
...
+
+// XXX: WIP -- I'm not sure how to properly test a failed connection.
+// At the moment, I'm thinking I'll just mark this test as "don't run"
+// and use it by hand for testing (which basically means run it, kill my
+// internet, start my internet again, watch what happens)
+selftest.define("resume downloads", function () {
+  const url = 'http://warehouse.meteor.com/builds/Pr7L8f6PqXyqNJJn4/1443478653127/aRiirNrp4v/meteor-tool-1.1.9-os.osx.x86_64+web.browser+web.cordova.tgz';
+
+  const result = require('../utils/http-helpers').getUrlWithResuming({
+    timeout: 1000,
```
Comment (benjamn) [inline tools/tests/utils-tests.js]: This should actually test resuming, perhaps by supporting a custom `options.outputStream` object, and passing a test object whose `.write` method throws an HTTP error after the first few chunks are written?

File: tools/utils/http-helpers.js [hunk 5]
```diff
@@ -14,6 +14,7 @@ var release = require('../packaging/release.js');
 var Console = require('../console/console.js').Console;
 var timeoutScaleFactor = require('./utils.js').timeoutScaleFactor;
 
+import { WritableStreamBuffer } from 'stream-buffers';
```
Comment (benjamn) [inline tools/utils/http-helpers.js]: The `stream-buffers` package needs to be in `dev_bundle/lib/node_modules`.

File: tools/utils/http-helpers.js [hunk 6]
```diff
@@ -358,6 +369,47 @@ _.extend(exports, {
...
+      }
+    }
+    const response = attempt(MAX_ATTEMPTS).response;
+    const body = outputStream.getContents();
+
+    if (response.statusCode >= 400 && response.statusCode < 600) {
+      const href = response.request.href;
+      throw Error(
+         body ||
+          `Could not get ${href}; server returned [${response.statusCode}]`);
```
Comment (benjamn) [inline tools/utils/http-helpers.js]: Maybe this was just for debugging, but I would use the same error message every time, regardless of whether `body` is defined, and then just `return outputStream.getContents()` below, if there was no error.

File: tools/utils/http-helpers.js (hunk 5 above)
Comment (benjamn) [inline tools/utils/http-helpers.js]: Ah, sorry, I see that in your to-do list now.

**Comment by benjamn** [comment]
I like the idea of doing this in memory!

File: tools/utils/http-helpers.js (hunk 3 above)
Comment (benjamn) [inline tools/utils/http-helpers.js]: Perhaps there should be a delay of a few seconds between attempts?

File: tools/utils/http-helpers.js (hunk 1 above)
Comment (benjamn) [inline tools/utils/http-helpers.js]: If you use `options` here, the `options.headers` property will override the `headers` property, so you might as well update `options.headers` above, and remove the `headers,` line here.

File: tools/utils/http-helpers.js [hunk 7]
```diff
@@ -358,6 +369,47 @@ _.extend(exports, {
...
+    const options = _.isObject(urlOrOptions) ? urlOrOptions : {
+      url: urlOrOptions,
+    };
+
+    const outputStream = new WritableStreamBuffer();
+
+    const MAX_ATTEMPTS = 5;
+    function attempt(triesRemaining) {
+      const headers = _.extend({}, urlOrOptions.headers, {
+        range: `bytes=${outputStream.size()}-`,
```
Comment (benjamn) [inline tools/utils/http-helpers.js]: Should `range` be capitalized here?

File: tools/tests/utils-tests.js (hunk 4 above)
Comment (tmeasday) [inline tools/tests/utils-tests.js]: Oooh, that might work, let me try it.

File: tools/utils/http-helpers.js (hunk 5 above)
Comment (tmeasday) [inline tools/utils/http-helpers.js]: Yeah I just wanted to check with you that the concept of using a stream buffer to collect the body made sense before bothering to build a new bundle etc.

File: tools/utils/http-helpers.js (hunk 3 above)
Comment (tmeasday) [inline tools/utils/http-helpers.js]: That's part of the plan, yeah.

File: tools/utils/http-helpers.js (hunk 6 above)
Comment (tmeasday) [inline tools/utils/http-helpers.js]: Yeah, I was just re-implementing what `getUrl` does above, but I agree it's sort of weird.

File: tools/utils/http-helpers.js (hunk 7 above)
Comment (tmeasday) [inline tools/utils/http-helpers.js]: :+1:

File: tools/utils/http-helpers.js (hunk 2 above)
Comment (tmeasday) [inline tools/utils/http-helpers.js]: :+1:

File: tools/utils/http-helpers.js (hunk 1 above)
Comment (tmeasday) [inline tools/utils/http-helpers.js]: 👍

**Comment by tmeasday** [comment]
@ben I think this should be good to go, awaiting tests

File: tools/packaging/tropohouse.js [hunk 8]
```diff
@@ -304,7 +304,7 @@ _.extend(exports.Tropohouse.prototype, {
     // it relies on extractTarGz being fast and not reporting any progress.
```
Comment (abernix) [inline tools/packaging/tropohouse.js]: It's just off the top of the diff here, but I think this `XXX/TODO` would be a huge benefit as it's getting hard to debug when something is stuck `downloading` or something is stuck `extracting`.

I'm getting the feeling that people are getting stuck _after_ the actual download, especially on Windows.

**Pull request merged**